tutorials/iot-device-to-device/index.md

```diff
@@ -0,0 +1,254 @@
+```
+
+To setup the Google Cloud Function, deploy it to trigger on the topic
+configured with your device registry.
+
+```
+    gcloud beta functions deploy relayCloudIot \
+        --stage-bucket=gs://your-gcs-bucket \
+        --trigger-topic=your-topic-id
+```
+
+## Register a Device
+
+Before you can connect a device, you must register its public key with the
+Google Cloud IoT Core device manager. There are a number of ways to do this,
+but for now, we'll use the terminal and `gcloud`.
+
+The following command will generate a RSA-256 keypair:
+
+```
+    openssl req -x509 -newkey rsa:2048 -days 3650 -keyout rsa_private.pem -nodes -out \
+        rsa_cert.pem -subj "/CN=unused"
+```
+
+After you have generated the keypair, use the public key to register your
+device:
+
+```
+    gcloud beta iot devices create \
+        --registry=<your-registry-id> \
+        --region "europe-west1" \
+        --public-key path=rsa_cert.pem,type=rs256 <deviceId>
+```
+
+Now that you have registered a device, you can connect it using the virtual
+device provided in the sample.
+
+##  Connect your virtual device
+
+The virtual device is based entirely on the [NodeJS MQTT device sample](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/tree/master/iot/mqtt_example)
+with two small differences: the telemetry message that is sent is updated and
+the handler for receieving messages is different.
+
+The following code shows how the telemetry message is generated and sent:
+
+```
+    const payload = {
+      cloudRegion: argv.cloudRegion,
+      deviceId: argv.deviceId,
+      registryId: argv.registryId,
+      hops: messagesSent
+    };
+    console.log('Publishing message:', payload);
+    client.publish(mqttTopic, JSON.stringify(payload), { qos: 1 }, function (err) {
+      if (!err) {
+        shouldBackoff = false;
+        backoffTime = MINIMUM_BACKOFF_TIME;
+      }
+    });
+```
+
+The following code shows how the demo app handles configuration change
+messages:
+
+```
+    client.on('message', (topic, message, packet) => {
+      console.log('message received: ', Buffer.from(message, 'base64').toString('ascii'));
+      let payload = JSON.parse(Buffer.from(message, 'base64').toString('ascii'));
+      console.log(`${payload.hops} to ${++payload.hops}`);
+      publishAsync(payload.hops, payload.hops+1);
+    });
+```
+
+To run the demo after you have successfully setup your Device Registry and
+deployed your Google Cloud Function, run the following command from the
+virtualdevice folder:
+
+```
+    cd virtualdevice
+    npm install
+    node virtualdev.js --cloudRegion=<region-from-console> \
+        --projectId<your-project-id> \
+        --deviceId=<your-device-id>\
+        --privateKeyFile=../rsa_private.pem \
+        --algorithm=RS256 \
+        --numMessages=1 \
+        --registryId=<your-registry-id>
+```
+
+When you connect your device, it will transmit a telemetry message that will
```

Comment:
In an ideal world, this description wouldn't be in the future tense. E.g. "When you connect your device, it transmits a telemetry message..." Not a show-stopper by any means.